### PR TITLE
Load leases into the expiration manager in parallel 

### LINKED ISSUE #2340: Parallel/lazy-load leases on startup
We should have the expiration manager lazy load leases on startup, and should parallelize the operation to load them faster.

### PR DESCRIPTION
**Pull request by briankassouf**
fixes #2340

### CODE REVIEW

File: vault/expiration.go [hunk 1]
```diff
@@ -125,46 +128,111 @@ func (m *ExpirationManager) Restore() error {
...
+			})
+		}
 	}
 	if len(m.pending) > 0 {
 		if m.logger.IsInfo() {
 			m.logger.Info("expire: leases restored", "restored_lease_count", len(m.pending))
 		}
 	}
+
+	wg.Wait()
```
Comment (vishalnayak) [inline vault/expiration.go]: Can we do the `Wait()` before the `if` check above?

File: vault/expiration.go [hunk 2]
```diff
@@ -125,46 +128,111 @@ func (m *ExpirationManager) Restore() error {
...
+			default:
+				broker <- leaseID
+			}
 		}
-		// Load the entry
-		le, err := m.loadEntry(leaseID)
-		if err != nil {
+
+		// Close the broker, causing worker routines to exit
+		close(broker)
```
Comment (vishalnayak) [inline vault/expiration.go]: Please correct me if I am wrong, won't this terminate the workers prematurely, immediately after assigning them leaseIDs to load?

File: physical/consul.go [hunk 3]
```diff
@@ -154,6 +154,10 @@ func newConsulBackend(conf map[string]string, logger log.Logger) (Backend, error
 
 	// Configure the client
 	consulConf := api.DefaultConfig()
+	// Set MaxIdleConnsPerHost to the number of processes used in expiration.Restore
+	tr := cleanhttp.DefaultPooledTransport()
+	tr.MaxIdleConnsPerHost = 64
```
Comment (briankassouf) [inline physical/consul.go]: Currently they are in two different packages with cyclic imports, so I can't use `RestoreWorkerCount`. let me think about how to make this a bit better.

File: vault/expiration.go (hunk 2 above)
Comment (briankassouf) [inline vault/expiration.go]: Because broker is not buffered, by the time the execution get here all the leases should have been loaded.

Comment (vishalnayak) [inline vault/expiration.go]: Ah, yes!

File: physical/consul.go (hunk 3 above)
Comment (jefferai) [inline physical/consul.go]: I've run into this problem and currently have been putting values into `logical` but doesn't belong there.

The right answer might be to create a `helper/const` package that literally just holds const values (that should be exported/shared). Feel free to do that here and I'll port some things over...

**Comment by jefferai** [comment]
Looks really good!

File: physical/consul.go (hunk 3 above)
Comment (briankassouf) [inline physical/consul.go]: 👍 I'll add that

**Comment by jefferai** [comment]
Pedantic: `const` vs `consts`? In conversation the latter is super awkward although it's obviously more correct. Just bringing that up, go with your gut.

**Comment by jefferai** [comment]
Actually, nm. You can overload built-in identifiers but it sucks. Keep `consts` :-)

**Pull request merged**